lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,1314 @@
+	// to test if the constructor would catch inputs/outputs that are attempting to be added
+	// with duplicate serial ids.
+	struct DuplicateEntropySource;
+	impl EntropySource for DuplicateEntropySource {
+		fn get_secure_random_bytes(&self) -> [u8; 32] {
+			let mut res = [0u8; 32];
+			let count = 1u64;
+			res[0..8].copy_from_slice(&count.to_be_bytes());
+			res
+		}
+	}
+
+	struct TestSession {
+		inputs_a: Vec<(TxIn, TransactionU16LenLimited)>,
+		outputs_a: Vec<TxOut>,
+		inputs_b: Vec<(TxIn, TransactionU16LenLimited)>,
+		outputs_b: Vec<TxOut>,
+		expect_error: Option<AbortReason>,
+	}
+
+	fn do_test_interactive_tx_constructor(session: TestSession) {
+		let entropy_source = TestEntropySource(AtomicCounter::new());
+		do_test_interactive_tx_constructor_internal(session, &&entropy_source);
+	}
+
+	fn do_test_interactive_tx_constructor_with_entropy_source<ES: Deref>(
+		session: TestSession, entropy_source: ES,
+	) where
+		ES::Target: EntropySource,
+	{
+		do_test_interactive_tx_constructor_internal(session, &entropy_source);
+	}
+
+	fn do_test_interactive_tx_constructor_internal<ES: Deref>(
+		session: TestSession, entropy_source: &ES,
+	) where
+		ES::Target: EntropySource,
+	{
+		let channel_id = ChannelId(entropy_source.get_secure_random_bytes());
+		let tx_locktime = AbsoluteLockTime::from_height(1337).unwrap();
+
+		let (mut constructor_a, first_message_a) = InteractiveTxConstructor::new(
+			entropy_source,
+			channel_id,
+			FEERATE_FLOOR_SATS_PER_KW * 10,
+			true,
+			tx_locktime,
+			session.inputs_a,
+			session.outputs_a,
+			0,
+		);
+		let (mut constructor_b, first_message_b) = InteractiveTxConstructor::new(
+			entropy_source,
+			channel_id,
+			FEERATE_FLOOR_SATS_PER_KW * 10,
+			false,
+			tx_locktime,
+			session.inputs_b,
+			session.outputs_b,
+			0,
+		);
+
+		let handle_message_send =
+			|msg: InteractiveTxMessageSend, for_constructor: &mut InteractiveTxConstructor| {
+				match msg {
+					InteractiveTxMessageSend::TxAddInput(msg) => for_constructor
+						.handle_tx_add_input(&msg)
+						.map(|msg_send| (Some(msg_send), None)),
+					InteractiveTxMessageSend::TxAddOutput(msg) => for_constructor
+						.handle_tx_add_output(&msg)
+						.map(|msg_send| (Some(msg_send), None)),
+					InteractiveTxMessageSend::TxComplete(msg) => {
+						for_constructor.handle_tx_complete(&msg)
+					},
+				}
+			};
+
+		assert!(first_message_b.is_none());
+		let mut message_send_a = first_message_a;
+		let mut message_send_b = None;
+		let mut final_tx_a = None;
+		let mut final_tx_b = None;
+		while final_tx_a.is_none() || final_tx_b.is_none() {
+			if let Some(message_send_a) = message_send_a.take() {
+				match handle_message_send(message_send_a, &mut constructor_b) {
+					Ok((msg_send, final_tx)) => {
+						message_send_b = msg_send;
+						final_tx_b = final_tx;
+					},
+					Err(abort_reason) => {
+						assert_eq!(Some(abort_reason), session.expect_error);
```

Comment:
Should we check that there are no more messages left to be sent? Otherwise, the error may occur at an unexpected time but not cause the test to to fail. Maybe a `break` would suffice if you `take` the `expect_error`?